[Merged by Bors] - beacon sync: wait for N blocks before deciding beacon value from blocks

countvonzero opened the pull request:
## Motivation
<!-- Please mention the issue fixed by this PR or detailed motivation -->
Closes #2810
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes
<!-- Please describe in detail the changes made -->
- for nodes that did not participate in the tortoise beacon in the previous epoch, wait for N blocks to decide the beacon value for the previous epoch
- node: wait for beacon value to be determined for previous epoch before participating in hare
- miner: wait for beacon value to be determined for previous epoch before producing blocks

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->
make the late-node test wait one more epoch due to the nature of beacon sync
- before:
late node participates in hare as soon as it is synced with a fake beacon value (epoch ID)
i.e. a late node will join at epoch 4, become synced at epoch 6 and consistently participates in hare during epoch 7.

- after:
late node only participates in hare once it learned of the beacon value from previous epoch. so it will only consistently participate in hare the epoch after it take parts in the tortoise beacon.
i.e. a late node will join at epoch 4, become synced and published an ATX in epoch 6, and participates in beacon protocol in epoch 7. so it only consistently participates in hare in epoch 8. 
after it gets synced, it will participate in hare in the 2nd and 3rd layer of the epoch 6/7 (after it syncs the beacon value during the 1st layer of epoch 6/7)


## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

countvonzero committed:
collect beacon values from blocks

countvonzero committed:
miner only produce block where beacon is available

countvonzero committed:
hare wait for beacon before participate

countvonzero committed:
update configs

countvonzero committed:
remove naive beacon sync code

countvonzero committed:
wait extra epoch for late node system test

countvonzero commented:
bors try

countvonzero commented:
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/runs/3750271469)

https://github.com/spacemeshos/go-spacemesh/runs/3750591105

unit tests was canceled. retry the jobs and all the system tests passed

dshulyak commented on cmd/root.go:
```diff
@@ -205,6 +205,8 @@ func AddCommands(cmd *cobra.Command) {
 		"Ratio of votes for reaching consensus")
 	cmd.PersistentFlags().Uint64Var(&config.TortoiseBeacon.VotesLimit, "tortoise-beacon-votes-limit",
 		config.TortoiseBeacon.VotesLimit, "Maximum allowed number of votes to be sent")
+	cmd.PersistentFlags().Uint32Var(&config.TortoiseBeacon.BeaconSyncNumBlocks, "tortoise-beacon-sync-num-blocks",
```
i am not sure that this is supposed to be configured separately. shouldn't it be based on the layer size and/or number of layers in epoch?

countvonzero replied:
according to Tal during the R&D meeting, he was not sure whether we should set it to 1, 1.5 or 2 clusters of (800) blocks, as 800 is the security param. it's likely we will have to experiment with this number. i.e. either set it to 1X/1.5X/2X or set the number of blocks.

also, for system testing, i am not sure whether 800 gets reduced to a smaller number. so it's also easier to explicitly set it to a small number, disregarding layer_size and num_layers_per_epoch.

i imagine when we get to a stage where we are sure of the network param, we can actually simplify a lot more.

dshulyak replied:
to me this number seems a bit magical, I've seen it mentioned in tortoise weakcoin as well. do you know if there is any dependency between layer size and layers per epoch? or how it was invented?

countvonzero replied:
i am not familiar with the tortoise weak coin (for self-healing).
800 is our security parameter in the whitepaper. during the R&D discussion, Tal and Iddo always refer 800 as a cluster. if we see 800 of something, and if more than 2/3 of them agree, then our security assumption holds.

if we just sticks to 800, then i can hard code a value in the code that says beaconSyncNumBlocks := 800.
as far as i know, we only have hare committee size set to 800.

in the R&D meetings, Tal's idea was to wait K layers that satisfy 800 or 1600 blocks.
but layers are ticked. let's say the average blocks per layer is 200. then we wait minimum 4 layers.
however, i don't want to wait for 4 full layers if we collected 800 blocks earlier than 4 layers. that's why i decided to use numBlocks instead of layers.

there is no relation to layers per epoch that i know of.

dshulyak commented on tortoisebeacon/tortoise_beacon.go:
```diff
@@ -193,52 +191,121 @@ func (tb *TortoiseBeacon) Close() {
-		val, err := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID - 1)
-		if err == nil {
-			return val.Bytes(), nil
+	if epochBeacon := tb.findMostWeightedBeaconBeaconForEpoch(epoch); epochBeacon != nil {
```
BeaconBeacon?

countvonzero replied:
i was maybe hungry and thinking about bacon when i wrote this. 🥓

dshulyak reviewed (commented):
great approach, very intuitive.

in case if node skipped participation in tortoise beacon when it wasn't synced at the start of epoch, it just needs to wait for a sufficient number of blocks before enabling all depending protocols and it doesn't matter where it will get those blocks from (sync or gossip). right?

i am bit concerned that db was removed, if node was restarted right after it got a beacon it will have to wait for many blocks again. if the beacon was persisted it will need only to wait for 2 gossip layers, or am i missing something?

nkryuchkov commented on hare/eligibility/beacon_test.go:
```diff
@@ -18,40 +18,12 @@ func (mbp mockBeaconProvider) GetBeacon(types.EpochID) ([]byte, error) {
-	r.Equal(binary.LittleEndian.Uint32(beaconValue), val)
-	r.Equal(1, c.numGet)
-	r.Equal(1, c.numAdd)
+	encodedValue := uint32(67305985) // binary.LittleEndian.Uint32(beaconValue)
```
I think hex number would be more readable here
```suggestion
	encodedValue := uint32(0x4030201) // binary.LittleEndian.Uint32(beaconValue)
```

countvonzero replied:
you are right. done.

nkryuchkov reviewed (approved):
Great job! The code became much cleaner

countvonzero committed:
add beacon db to persist data

countvonzero committed:
every block can report beacon

countvonzero commented:
> in case if node skipped participation in tortoise beacon when it wasn't synced at the start of epoch, it just needs to wait for a sufficient number of blocks before enabling all depending protocols and it doesn't matter where it will get those blocks from (sync or gossip). right?

correct. the source of the data (sync/gossip) does not matter. however, each miner only records the beacon in their **first** block in the epoch. so i am thinking the blocks may be concentrated at the beginning of the epoch.

before i only recorded the block that actually has beacon value. it's trivial to record it from every block. i have changed the code to do so (see the last commit)
 
> i am bit concerned that db was removed, if node was restarted right after it got a beacon it will have to wait for many blocks again. if the beacon was persisted it will need only to wait for 2 gossip layers, or am i missing something?

you are right. i simply was willing to let the node wait longer. added a private db inside tortoise beacon to speed up the recovery.

countvonzero committed:
unflake parallel test: set the same layer per epoch in tortoisebeacon pkg

countvonzero committed:
reduce logging from dev testing

countvonzero committed:
limit number of epochs to keep

dshulyak commented on tortoisebeacon/tortoise_beacon.go:
```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+	if tb.db == nil {
+		return
+	}
+	if err := tb.db.Put(epoch.ToBytes(), beacon.Bytes()); err != nil {
```
if persistence is enabled this error should be checked, why to "optimize" this and make persistence best-effort?

countvonzero replied:
ok. i didn't think it's worth failing other component if there is error here. but if disk is failing, we have bigger issues. changed as you suggested

dshulyak commented on tortoisebeacon/tortoise_beacon.go:
```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+}
+
+func (tb *TortoiseBeacon) persistBeacon(epoch types.EpochID, beacon types.Hash32) {
+	if tb.db == nil {
```
but why to make this optional? i think it was optional in previous code as well, but i don't see any reason to do so

countvonzero replied:
my reason was for ease of testing. but i think i was tunnel visioned. thanks for pushing me. changed.

dshulyak commented on tortoisebeacon/tortoise_beacon.go:
```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+func (tb *TortoiseBeacon) getBeacon(epoch types.EpochID) []byte {
+	tb.mu.RLock()
+	defer tb.mu.RUnlock()
+	if beacon, ok := tb.beacons[epoch]; ok {
```
why not to store data only in db? it is not going to be a bottleneck

countvonzero replied:
it is not going to be a bottleneck for sure but i still don't like to access disk every time a beacon is queried.

our protocol is based on synchronous network. and some messages are time sensitive. the hardware assumption for home miners are **not** SSD but HDD. i think we should still avoid disk seeks and the long tail on disk read time as much as we can. the epoch numbers to keep is limited to 10. i'm not concerned abt memory usage.

dshulyak commented on tortoisebeacon/tortoise_beacon.go:
```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+	if tb.db == nil {
+		return nil
+	}
+	if beacon, err := tb.db.Get(epoch.ToBytes()); err == nil {
```
why not to return an error?

countvonzero replied:
done

dshulyak reviewed (approved):
i am not sure that it is good idea to make persistence best-effort, but up to you

countvonzero committed:
report error from persisting beacon

countvonzero commented:
> to me this number seems a bit magical, I've seen it mentioned in tortoise weakcoin as well. do you know if there is any dependency between layer size and layers per epoch? or how it was invented?

@dshulyak i thought the comment in tortoisebeacon/config.go is adequate.

i am not familiar with the tortoise weak coin (for self-healing).
800 is our security parameter in the whitepaper. during the R&D discussion, Tal and Iddo always refer 800 as a cluster. if we see 800 of something, and if more than 2/3 of them agree, then our security assumption holds.

if we just sticks to 800, then i can hard code a value in the code that says beaconSyncNumBlocks := 800.
as far as i know, we only have hare committee size set to 800.

in the R&D meetings, Tal's idea was to wait K layers that satisfy 800 or 1600 blocks. 
but layers are ticked. let's say the average blocks per layer is 200. then we wait minimum 4 layers.
however, i don't want to wait for 4 full layers if we collected 800 blocks earlier than 4 layers. that's why i decided to use numBlocks instead of layers.

there is no relation to layers per epoch that i know of.

countvonzero commented:
bors merge
